### ISSUE TITLE
iOS crash because of SIGSEGV(SEGV_ACCERR) GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer

### ISSUE THREAD

**Issue by cj3479** [flutter/flutter#153697]
### Steps to reproduce

Many users have these crash, The stack information is as follows because it's an online crash, I don't know how to reproduce it

### Expected results

no crash

### Actual results

crash

### Code sample

It is not a mandatory crash, so there is no sample code.

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

![image](https://github.com/user-attachments/assets/6d039475-3598-4335-8d4c-8b9f0a4ebf9f)

```console
Flutter 0x0000000113693804 std::_fl::__function::__func<fml::internal::CopyableLambda<flutter::GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer(SkISize const&)::$_0>, std::_fl::allocator<fml::internal::CopyableLambda<flutter::GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer(SkISize const&)::$_0> >, boolean (flutter::SurfaceFrame&, flutter::DlCanvas*)>::operator()(flutter::SurfaceFrame&, flutter::DlCanvas*&&) (../../third_party/libcxx/include/__memory/shared_ptr.h:815)

Flutter 0x000000011369378c std::_fl::__function::__func<fml::internal::CopyableLambda<flutter::GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer(SkISize const&)::$_0>, std::_fl::allocator<fml::internal::CopyableLambda<flutter::GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer(SkISize const&)::$_0> >, boolean (flutter::SurfaceFrame&, flutter::DlCanvas*)>::operator()(flutter::SurfaceFrame&, flutter::DlCanvas*&&) (../../third_party/libcxx/include/__memory/shared_ptr.h:701)


Flutter 0x000000011357fffc flutter::SurfaceFrame::Submit() (../../flutter/flow/surface_frame.cc:51)


Flutter 0x00000001134bbdd8 flutter::Rasterizer::DrawToSurfacesUnsafe(flutter::FrameTimingsRecorder&, std::_fl::vector<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask>>, std::_fl::allocator<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask>>>>) (../../flutter/shell/common/rasterizer.cc:752)


Flutter 0x00000001134bc2c4 std::_fl::__function::__func<flutter::Rasterizer::DrawToSurfaces(flutter::FrameTimingsRecorder&, std::_fl::vector<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask> >, std::_fl::allocator<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask> > > >)::$_1, std::_fl::allocator<flutter::Rasterizer::DrawToSurfaces(flutter::FrameTimingsRecorder&, std::_fl::vector<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask> >, std::_fl::allocator<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask> > > >)::$_1>, void ()>::operator()() (../../flutter/shell/common/rasterizer.cc:552)


Flutter 0x00000001133b32c4 fml::SyncSwitch::Execute(fml::SyncSwitch::Handlers const&) const (../../flutter/fml/synchronization/shared_mutex.h:29)



Flutter 0x00000001134bb510 flutter::Rasterizer::DrawToSurfaces(flutter::FrameTimingsRecorder&, std::_fl::vector<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask>>, std::_fl::allocator<std::_fl::unique_ptr<flutter::LayerTreeTask, std::_fl::default_delete<flutter::LayerTreeTask>>>>) (../../third_party/libcxx/include/__functional/function.h:470)


Flutter 0x00000001134bd114 std::_fl::__function::__func<flutter::Rasterizer::Draw(std::_fl::shared_ptr<flutter::Pipeline<flutter::FrameItem> > const&)::$_0, std::_fl::allocator<flutter::Rasterizer::Draw(std::_fl::shared_ptr<flutter::Pipeline<flutter::FrameItem> > const&)::$_0>, void (std::_fl::unique_ptr<flutter::FrameItem, std::_fl::default_delete<flutter::FrameItem> >)>::operator()(std::_fl::unique_ptr<flutter::FrameItem, std::_fl::default_delete<flutter::FrameItem> >&&) (../../third_party/libcxx/include/vector:423)


Flutter 0x00000001134bc960 flutter::Rasterizer::Draw(std::_fl::shared_ptr<flutter::Pipeline<flutter::FrameItem>> const&) (../../third_party/libcxx/include/__memory/unique_ptr.h:259)

Flutter 0x00000001134d4e14 std::_fl::__function::__func<fml::internal::CopyableLambda<flutter::Shell::OnAnimatorDraw(std::_fl::shared_ptr<flutter::Pipeline<flutter::FrameItem> >)::$_0>, std::_fl::allocator<fml::internal::CopyableLambda<flutter::Shell::OnAnimatorDraw(std::_fl::shared_ptr<flutter::Pipeline<flutter::FrameItem> >)::$_0> >, void ()>::operator()() (../../flutter/shell/common/shell.cc:1260)

Flutter 0x00000001133b1930 fml::MessageLoopImpl::FlushTasks(fml::FlushType) (../../flutter/fml/message_loop_impl.cc:128)

Flutter 0x00000001133b50e4 fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) (../../flutter/fml/platform/darwin/message_loop_darwin.mm:86)
```

### Flutter Doctor output

```console
[!] Flutter (Channel [user-branch], 3.19.6, on macOS 14.3 23D56 darwin-x64, locale zh-Hans-CN)
! Flutter version 3.19.6 on channel [user-branch] 
Currently on an unknown channel. Run flutter channel to switch to an official channel.
If that doesn't fix the issue, reinstall Flutter by following instructions at https://flutter.dev/docs/get-started/install.
! Upstream repository unknown source is not a standard remote.
Set environment variable "FLUTTER_GIT_URL" to unknown source to dismiss this error.
• Framework revision https://github.com/flutter/flutter/commit/54e66469a933b60ddf175f858f82eaeb97e48c8d (2 个月前), 2024-04-17 13:08:03 -0700
• Engine revision c4cd48e186
• Dart version 3.3.4
• DevTools version 2.31.1
• If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades
```

**Comment by huycozy** [comment]
The crash trace is similar to https://github.com/flutter/flutter/issues/149885 and it's confirmed as fixed on Flutter 3.22.2 and enabling impeller. Closing as a duplicate.

**Comment by cj3479** [comment]
> ```
> GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer
> ```

Can you see the crash log "GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer"  which is different  from "GPUSurfaceMetalSkia::AcquireFrameFromCAMetalLayer"  carefully?
Obviously, The impeller occurs crash and I am very sure, I enable impeller

**Comment by cj3479** [comment]
> The crash trace is similar to #149885 and it's confirmed as fixed on Flutter 3.22.2 and enabling impeller. Closing as a duplicate.

Can you find the crash log "GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer" which is different from "GPUSurfaceMetalSkia::AcquireFrameFromCAMetalLayer" ?
Obviously, The impeller occurs crash and I am very sure, I enable impeller

**Comment by huycozy** [comment]
Please see the updated trace from the author of that issue here https://github.com/flutter/flutter/issues/149885#issuecomment-2154669673 which also contains `GPUSurfaceMetalImpeller::AcquireFrameFromCAMetalLayer`. And please try upgrading Flutter SDK to 3.22.2+.
